### ISSUE TITLE
wrong accumulated altitude of recorded gpx track

### ISSUE THREAD

**Issue by wanderfalk** [osmandapp/Osmand#4481]
recording a gpx track, the accumulated altitude is not calculated correctly.

On a couple of hikes of some 20km in mid range mountains several 1000 meters are displayed in dashboard->tracks->current track->height->ascending/decending
I would expect roughly one tenth of that value

reproduced with several of the latest osmand nightly builds on moto g4, android 7.0
example: 10kms, osmand +3745m -3662m
[altitude_problem.txt](https://github.com/osmandapp/Osmand/files/1330868/altitude_problem.txt)
NB: 
-speed and time lines removed
-garmin basecamp calculates +1022m -1027m from this file

**Comment by vshcherb** [comment]
[altitude_csv.txt](https://github.com/osmandapp/Osmand/files/1347053/altitude_csv.txt)
To simplify investigation here is the csv file:
lat,lon,distance_to_previous_point,altitude. So analysis could be easily done in Excel

**Comment by vshcherb** [comment]
Simple test gives these result:
0 meter distance filter:  +4650 -4598
1 meter distance filter:  +3567 -3589
5 meter distance filter:  +2732 -2577
7 meter distance filter:  +1743 -1688

**Comment by vshcherb** [comment]
Slightly changed formula. We accumulate distance until threshold T and only after that evaluate asc/desc & end to the sum
0 meter distance filter: +4650 -4598
1 meter distance filter: +3740 -3690
5 meter distance filter: +3124 -2954
7 meter distance filter: +2353 -2399
10 meter distance filter: +1822 -1801
20 meter distance filter: +1250 -1181

**Comment by vshcherb** [comment]
Implementing different algorithm - smoothing data by distance i.e.
calculate Average(height) every N meters. 
For example, 0 m - 100m, 1 m - 105m, 2 m - 101m, 3m - 105m, 4m - 109m. If we smooth data by 2 meters, than height array will be  [100m, 103m, 107m] - total asc is 7 meters vs 13 meters.
Using this algorithm data is:
0 meter distance filter: +4650 -4598
1 meter distance filter: +3737 -3744
5 meter distance filter: +3029 -3036
7 meter distance filter: +2341 -2348
10 meter distance filter: +1789 -1791
20 meter distance filter: +1101 -1102

**Comment by sonora** [comment]
Current approach is dynamic channel based, much like you use for stock market analysis, plus factoring in the accuracy of your GPS reception. Let me analyze your track when I will be back from travels (will take a few weeks). But at the time I did the coding, I think OsmAnd's approach was rather innovative compared to most other GPS software, which either performed simple accumulation over all measurement errors (overestimating), or applied general smoothing which did not apply to all use cases (underestimating for e.g. hikers).

**Comment by vshcherb** [comment]
Well dynamic channel approach certainly has more disadvantages comparing to static analysis approach. Though my conclusion that 1 parameter basically shows more or less same results regardless of approach. All dynamic approaches also have that parameter called *window* let's see if it gives different results.

**Comment by sonora** [comment]
Victor, this is completely correct: It boils down to one parameter, pretty much regardless of the method applied. And the purpose of that parameter is to filter the noise from the signal.
Unfortunatrly elevation noise is significantly higher than lateral noise in practice, which can easily be verified by going through recordings taken under varying circumstances.
That is also why many methods tend to overestimate elevation gain/loss in the eyes of many users. But I am rather confident there is a good way to optimize the output for common use cases without additional user setting.

**Comment by vshcherb** [comment]
@sonora According to my rough analysis, we are using some kind of "window" of a length *1 m*, 3 methods that I used are giving 4000 meters with *1 meter window*. Does it make sense to increase window to *5 meters* or *7 meters*. 
For me it looks more realistic to calculate average height every *7 meters* and then calculate asc/desc using these average heights.

I'm ready to help to implement some kind of this algorith (I think we did something similar for slope graphics).

**Comment by sonora** [comment]
Victor, currently out traveling, cannot look at any code. From what I remember, my old channel approach does not throw away any values nor perform any smoothing laterally, instead simply tries to detect up/down directional (channel) changes before summing up respective portions. This also caters for climbers with near vertical elevation changes (this may be a rare use case). But of course such an approach is very sensitive to singular noise, like many bad GPS units exhibit, just like the track supplied above with many singular artefacts of +/-50m or so. I think that in deed a smoothing approach may give better values for most use cases for tgese vad GPS receivers (except where people only take very few readings). If you want you can go ahead and use the approach we used for the graphics display. I can do a little scientific research in a few weeks and see if I can improve upon it. ;)

**Comment by sonora** [comment]
Some scientific thoughts about this:

When comparing different appliances and web sites, it looks like all of them come up with vastly different values for total ascent/descent for the same track. Some even for different tracks recorded while running the same trip repeatedly. This seems to indicate there is no generally accepted conventional method to provide good accumulated ascent/descent figures.

I am even afraiday there may be no one good numerical value at all: User complaints about this very topic are extremely common in all related forums. Looks like in principle this issue is related to the Coastline Paradox https://en.m.wikipedia.org/wiki/Coastline_paradox. 

A viable solution approach is we apply a Low Pass Filter (LPF) to the measured data. The cut off frequency should best be derived from (or at least be related to) the speed values if provided in the data itself (essentially saying that any elevation changes faster than the actual speed must be bogus), or where speed is not available maybe really falling back to speed derived from lateral motion. Should this not suffice, simply apply a lateral filter directly, i.e. x meters, the route you were going. I am still hoping we can somehow find a useful parametrization much like is used for stock market (channel) analysis. But I think we will have to bury hopes that there is the one 'good' value or method, and when implementing it everyone would happily accept the results. happy.

Since all tools I checked report totally different values, I guess experienced hikers will never rely on this numerical ascent figure too much anyway, what counts more in practice is looking at the graph representation of the elevation profile, which we have anyway.

So in conclusion your suggestion may be very valid that we simply use the same smoothing we apply for the graphs and also derive our ascent/descent value from it. The one drawback is if we apply the same smoothing to graphs and numerical value calculation, we have lost in OsmAnd all indication how noisy a recording really is. But I guess for the common user the benefit outweighs this 'expert' drawback.

**Comment by sonora** [comment]
PS: This may be worth a shot: http://blog.mtbguru.com/2010/01/12/elevation-accuracy-revisited-2/

**Comment by sonora** [comment]
The test track provided above is extremely problematic, see the artefacts in this screenshot of what I think are clustered bogus variations of over 50m . I doubt issues like this can usefully be entirely solved, but see my suggestion below.

![screenshot_2017-10-09-08-26-18](https://user-images.githubusercontent.com/845735/31326985-3fc9dd52-accc-11e7-8396-e92dee15977c.png)

**Comment by sonora** [comment]
My current favorite is this approach: To create the ascent/descent accumulation let us do the following:

1. We apply an LPF to the elevation data, a simple linear smoothing by a running average of the last 5 points may suffice

2. To this data we apply my channel based approach to detect the significant directional changes (turnaround points), i.e. detect when the fluctuation channel changes the up/down direction for more than what we consider a significant elevation delta threshold (could be 5 or 10m or so)

3. For the summation we only take elevation data between the beginning and end section points of each of these directional channel segments.

From what I recall this may be pretty much my original implementation only that I do not apply the LPF smoothing I now describe in step 1, so @vshcherb perhaps this is an easy experiment from our current code?

**Comment by vshcherb** [comment]
@sonora no worries it could wait when you are back. Again it looks like smart and very smart algorithm provide very similar results based on "FILTER PARAMETER". I would like to change our parameter that we will get something similar to 7 Meters (what I investigated)
7 meter distance filter: +1743 -1688

I think this is the most reasonable value

**Comment by sonora** [comment]
Ok, will submit my refinement in November, and we can test where this leads. Still hoping I can somehow relate lateral hdop to the vertical fluctuation, in theory this should be related via a factor of 2-3.

Just curious: When you say '7 meter distance filter', are the 7 meters a lateral or a vertical threshold? Because I think in order to cover as many use cases as possible we should use vertical threshold which we consider to seperate true trend reversal  from fluctuation. If we use lateral parameters there would be big dependencies on car/bike/ped use cases and on sampling rate (recording interval chosen by user).

**Comment by vshcherb** [comment]
It is a lateral threshold. We shouldn't use vertical cause we don't have any assumption how vertical could change, though we know that it is very unlikely for most of our tracks have hills up/down every 5 steps.

**Comment by sonora** [comment]
Ok, implemented (https://github.com/osmandapp/Osmand/commit/b33c87bf03a1758e0d15b0b506627a49d7d4ee2c). I solved it in this way:

**(1)** For the Ascent/Descent accumulation, first apply a low pass filter (LPF) to the raw elevation data, this smoothes the vertical HF noise. Easiest achieved by a running average of say 5 measurement values. (More smoothing could be achieved by increasing to 7 or 9.)

**(2)** I changed the (vertical) turnaround detection for climb/descend and vice versa from 5 to 10 meters. This value essentially defines the minimum we consider as a relevant elevation change (of the smoothed data), i.e. below is considered noise.

So by playing with the following 2 parameters one can tune things to what you consider relevant in any scenario. I find that
- **smoothWindow = 5 measurements points** sufficiently smoothes recorded data for most devices and reception conditions
- **channelThresMin = 10 meters** seems appropriate for most ped/bicycle/car use cases

While things may vary, I think we should strive to find fix values for these 2 parameters, and leave variable analysis to more dedicated evaluation programs.

For all tracks I have checked so far, the above 5points/10meters set works extremely well. Also tested in a helicopter - works fine. The outcome for the rather flat and very noisy track provided in this issue originally is now 1142m/1143m. Only for extremely "flat" tracks, the threshold of now 10 meters means that we may now see a little too much filtering.

**Comment by vshcherb** [comment]
@sonora Good work. From my observations 10 meters was also the best outcome

**Comment by sonora** [comment]
Yeah. The beauty is I get along without any lateral threshold here, this is strictly a 1D vertical fluctuation analysis.
